### PR TITLE
Reduce repetition in Decimal binary kernels, upgrade to arrow 11.1

### PR DESCRIPTION
**Pull request by alamb**
# Rationale
I noticed some things that could be written "more nicely" while trolling through the code for other reasons and figured I would sneak in some coding (rather than review)

I think this also makes it easier to port this code to arrow-rs eventually: https://github.com/apache/arrow-rs/issues/1200

Changes:
1. Remove repetition in the code
2. Upgrade to arrow-rs `11.1` as it requires https://github.com/apache/arrow-rs/pull/1506

### CODE REVIEW

File: datafusion-physical-expr/src/expressions/binary.rs [hunk 1]
```diff
@@ -135,260 +135,192 @@ fn is_not_distinct_from_bool(
...
-    let mut bool_builder = BooleanBuilder::new(left.len());
-    for i in 0..left.len() {
-        if left.is_null(i) {
-            bool_builder.append_null()?;
-        } else {
-            bool_builder.append_value(left.value(i) == right)?;
-        }
-    }
-    Ok(bool_builder.finish())
+    compare_decimal_scalar(left, right, |left, right| left == right)
```
Comment (alamb) [inline datafusion-physical-expr/src/expressions/binary.rs:181]: This PR simply refactors out the common iteration over inputs into `compare_decimal` and `compare_decimal_scalar`

**Comment by alamb** [comment]
FYI @liukun4515

**Comment by alamb** [comment]
Draft until I can debug CI failure

**Comment by alamb** [comment]
Blocked by upstream issue in arrow-rs: https://github.com/apache/arrow-rs/pull/1506

File: datafusion-physical-expr/src/expressions/binary.rs [hunk 2]
```diff
@@ -135,260 +135,192 @@ fn is_not_distinct_from_bool(
...
+    left: &DecimalArray,
+    right: i128,
+    op: F,
+) -> Result<BooleanArray>
+where
+    F: Fn(i128, i128) -> bool,
+{
+    Ok(left
+        .iter()
+        .map(|left| left.map(|left| op(left, right)))
```
Comment (liukun4515) [inline datafusion-physical-expr/src/expressions/binary.rs:150]: Does this iter need `sized`？

Comment (alamb) [reply datafusion-physical-expr/src/expressions/binary.rs:150]: I think the call to `collect()` eventually needs the iterator to be sized (so it knows how large an array to allocate)

**Review by yjshen — APPROVED** [review]

**Pull request merged**